서보성 Seminar2 과제

Boseong-Seo opened the pull request:
(기능 전부 구현했으나 결과 사진 정리하는데 시간이 오래 걸려 PR 먼저 올립니다. 😢 ) -> 정리 완료
(11/1 오후 12:43) 프로필 수정 API 구현 코드 수정

---
# 🍒 회원가입

>  ### *구현 사항*
> - 회원가입에 성공하면, 유저는 AccessToken (JWT)를 응답으로 받아야
> - 회원가입 시 유저는 이메일(`email`), 이름(`username`), 비밀번호(`password`)를 입력
>    + 이때, 프로필을 함께 받을 수 있음.

--
### 🌵 INSTRUCTOR
> - 프로필에 소속 회사(`company`)와 몇 년차(`year`)인지를 저장할 수 있음. 
>   + default value : 각각 ""과 null으로 DB에 저장
<img width="1206" alt="스크린샷 2022-10-31 오후 11 42 31" src="https://user-images.githubusercontent.com/90292371/199035146-0abfa7d7-4f50-45d4-b725-01603ea4152b.png">
<img width="1162" alt="스크린샷 2022-10-31 오후 11 43 07" src="https://user-images.githubusercontent.com/90292371/199035301-dfa2da72-49b3-46ca-9622-421791ede330.png">

### 🌵 PARTICIPANT
> - 프로필에 소속 대학(`university`)과 활성회원 여부(`isRegistered`)를 저장할 수 있음.
>    + default value : 각각 `""`과 `true`로 DB에 저장
<img width="1113" alt="스크린샷 2022-11-01 오전 2 54 36" src="https://user-images.githubusercontent.com/90292371/199076113-a6b714f5-8613-4cb2-a8d4-d013e1e11673.png">

---
### ☑️ DB Result
위 요청에 따라 다음과 같이 DB에 회원 정보들이 저장된다.
앞서 기재한 요건(default value. etc..)에 맞게 값이 잘 들어간 듯 보인다.
위에서부터 순서대로, **Users**, **ParticipantProfile**, **InstructorProfile** 테이블에 해당한다.
<img width="1434" alt="스크린샷 2022-11-01 오전 2 58 47" src="https://user-images.githubusercontent.com/90292371/199076951-98b6344b-f2e0-4c95-a588-76ee81ddb1d4.png">

---
### 🚫 예외 케이스
#### 1) 이메일 중복 
: 409 Conflict Error
<img width="1113" alt="스크린샷 2022-11-01 오전 4 20 40" src="https://user-images.githubusercontent.com/90292371/199091944-54e4a29c-f9b0-49f7-8b71-49d7d45ecb85.png">

#### 2) 진행자 유저 가입, `year`에 0 이하 값 입력
: 400 Bad Request
<img width="1113" alt="스크린샷 2022-11-01 오전 3 04 44" src="https://user-images.githubusercontent.com/90292371/199078056-1c3d5879-f166-48f5-8da4-5da759de226a.png">


---
# 🍒 로그인

>  ### *구현 사항*
> - 로그인 시 유저는 이메일(`email`) & 비밀번호(`password`)를 입력하고, accessToken을 제공 받음
> - 로그인에 성공하면, **Users** 테이블의 해당 row의 마지막 로그인 시각(`lastLogin`) field 값도 그때 시각으로 수정되어야

아래는 user id가 1인 유저의 로그인 API 요청 결과이다.
<img width="1206" alt="스크린샷 2022-10-31 오후 11 45 41" src="https://user-images.githubusercontent.com/90292371/199035947-340e8833-e498-4833-ae03-ef87b74e16b4.png">

---
### ☑️ DB Result
로그인 성공으로, 위 유저에 해당하는 row의 `lastLogin` 컬럼이 최근 로그인 시각으로 변경된다.
(앞서 첨부한 DB Result 스크린샷을 비교해보면, 더 쉽게 확인이 가능하다.)
<br>
<img width="1263" alt="스크린샷 2022-11-01 오전 3 10 30" src="https://user-images.githubusercontent.com/90292371/199079174-89cf8400-fde0-4de5-84c9-8b8ad0c62f50.png">
<br>

---
### 🚫 예외 케이스
#### 1) 필수 입력값인 이메일 혹은 패스워드 입력 x
: 400 Bad Request
(여기서는 이메일을 입력하지 않는 케이스의 결과 스크린샷만 첨부하였다.)
<img width="1069" alt="스크린샷 2022-11-01 오전 3 16 46" src="https://user-images.githubusercontent.com/90292371/199080427-3066c966-310f-4c39-ace5-076dbe4cb849.png">
#### 2) 존재하지 않는 이메일
: 404 Not Found
<img width="1162" alt="스크린샷 2022-10-31 오후 11 46 20" src="https://user-images.githubusercontent.com/90292371/199036098-7c1444e4-8a20-4a3b-8614-4fe07a3cc022.png">
#### 3) 비밀번호 오류 
: 401 UnAuthorized
<img width="1206" alt="스크린샷 2022-10-31 오후 11 46 42" src="https://user-images.githubusercontent.com/90292371/199036192-d038dff3-8ccb-44eb-a0b3-b5c97fe71b1d.png">

---
# 🍒 프로필

### 🌵 프로필 확인
아래는 user id가 1인 유저의 프로필을 읽어온 결과이다.
<img width="1206" alt="스크린샷 2022-10-31 오후 11 47 32" src="https://user-images.githubusercontent.com/90292371/199036373-7214673c-ebb0-4b4a-96ee-f9d4c2795f2e.png">
이 유저에 해당하는 `ParticipantProfile`이 없어 요청 사항에 맞게 *participant*의 value는 null로,
해당 InstructorProfile은 있으나 아직 담당하는 세미나가 없으므로 *instructingSeminars*의 value 또한 null로 나온다.
<br>
이 유저에게 수강 권한을 부여한 후, 다시 한번 프로필 확인을 해보면은 아래와 같이 나오게 된다.
<img width="1053" alt="스크린샷 2022-11-01 오후 2 45 41" src="https://user-images.githubusercontent.com/90292371/199167476-3b600ad0-c502-4609-8ec6-769d0d00d1a0.png">

---
### 🌵 프로필 수정 (INSTRUCTOR)
>  #### *구현 사항*
> - 소속 회사(`company`)와 본인의 경력(`year`) 정보를 수정할 수 있음.
>   + default value : 각각 ""과 null으로 DB에 저장
>   + `company`에 0글자가 오는 경우(공백 등으로만 구성된 문자열)에는 ""로 수정
>   + `year`는 양수 값으로만 수정 가능.
>        - 0 이하의 값이 들어오면 400 Bad Request 에러
>   + 본인 유형과 맞지 않는 정보(=소속 학교(`university))가 request body에 함께 들어온다면, 무시
> - 위 정보와 함께, 이메일을 제외한 본인의 모든 정보를 수정할 수 있어야.
> - request body가 비어있다면, 무시

이번에는 user id가 2인 유저의 프로필 수정 요청을 시도해 보았다.
우선, 단순히 `username`만 수정해보았다. 
<img width="1206" alt="스크린샷 2022-10-31 오후 11 48 59" src="https://user-images.githubusercontent.com/90292371/199036749-9893ab3c-9d73-4cf5-a1d4-630776640ece.png">

그런 다음, username과 company 값까지 수정 요청을 보내보았다.
이때 company의 값을 `"      "`으로 수정하고자 하였다. 구현 의도대로라면 이 유저의 company 값에는 `""`이 들어가야 하는데, 본 요청에 대한 response body를 보면 그렇게 잘 반영된 것을 확인해볼 수 있다.
<img width="1113" alt="스크린샷 2022-11-01 오전 11 51 56" src="https://user-images.githubusercontent.com/90292371/199148232-7b8c23a0-455f-4abb-981b-5e0b9ae7df43.png">

---
### ☑️ DB Result
실제로 DB 상에도 잘 적용되었는지도 확인해보자.

1. username만 변경한 케이스
: user id 2에 해당하는 row의 `username`이 요청대로 잘 변경되었다.
<img width="1261" alt="스크린샷 2022-11-01 오전 3 38 05" src="https://user-images.githubusercontent.com/90292371/199084222-a1588643-b339-4ec0-a70c-1e4b39225b94.png">
<br>
2. username과 company 까지 수정
: user id 2인 유저의 InstructorProfile 정보에 `company` 수정 요청이 잘 반영되었음을 확인해볼 수 있다.
<img width="1508" alt="스크린샷 2022-11-01 오전 11 54 10" src="https://user-images.githubusercontent.com/90292371/199148554-c9d6ae6f-fa12-47c1-851e-cea78887e1e2.png">

이처럼 프로필 수정 내용이 DB에도 잘 반영된 것을 확인해볼 수 있다.

---
### 🌵 프로필 수정 (PARTICIPANT)
>  #### *구현 사항*
> - 진행자 User가 수강 권한을 부여받을 수 있음에 따라, 수정 API를 보내는 유저가 ParticipantProfile을 가지고 있는지 여부에 따라 요청을 수행하도록 구현.
> - 소속 대학(university) 수정 가능. 단, 활성회원 여부는 수정 불가
>   + default vale : `""`로 DB에 저장
>   + `university`에 0글자가 오는 경우(공백 등으로만 구성된 문자열)에는 ""로 수정

- 소속 대학 수정
user id가 3인 유저에 대해 `university` 값을 수정하는 API 요청을 보냄
<img width="1113" alt="스크린샷 2022-11-01 오후 12 21 25" src="https://user-images.githubusercontent.com/90292371/199151620-fa70cd3a-b5c8-4e48-81d9-11ba7deb25b8.png">

### ☑️ DB Result
위 API 요청에 대한 수행 결과가 성공적으로 DB에도 반영되었음을 확인해볼 수 있다.
<img width="1508" alt="스크린샷 2022-11-01 오후 12 22 04" src="https://user-images.githubusercontent.com/90292371/199151699-7e9462e0-4d01-4202-8992-43f3c02eb450.png">


--
### 🚫 예외 케이스
#### 1) 이메일 수정 시도 (두 유형 공통 케이스)
: 403 Forbidden
request body에 이메일 값이 들어오는 경우 그냥 무시하는 방법도 있겠지만, 경고의 의미로 에러를 띄워주는 것이 더 바람직할 거로 판단하였다. 이같은 이유로 이같이 구현.
<img width="1206" alt="스크린샷 2022-10-31 오후 11 49 39" src="https://user-images.githubusercontent.com/90292371/199036912-72dd3036-ed80-44dd-816f-e739f9a61eef.png">

#### 2) `year` 값을 0 이하의 값으로 변경 시도 (진행자 유형에만 해당)
: 400 Bad Request
<img width="1069" alt="스크린샷 2022-11-01 오후 12 02 39" src="https://user-images.githubusercontent.com/90292371/199149577-57562585-ce81-44b2-89c6-bbf2366a9194.png">

#### 2) 활성회원 여부(`isRegistered`) 수정 시도 (수강 권한이 있는 유저 해당)
: 403 Forbidden
request body에 해당 필드에 대한 값이 들어오는 경우 그냥 무시하는 방법도 있겠지만, 앞서 이메일의 경우와 같은 이유로 403 에러를 띄우도록 구현. 
<img width="1113" alt="스크린샷 2022-11-01 오후 12 22 58" src="https://user-images.githubusercontent.com/90292371/199151769-8eb791ec-d2ea-421f-bdb8-924f0da108ce.png">


---
# 🍒 수강 권한 요청
## ft. 강사도 수강생이 하고 싶다!

>  ### *구현 사항*
> - 기존에 진행자로 가입한 경우, 나중에 참여자로 등록 가능
>    + 즉, 유저의 role이 `INSTRUCTOR`이고 `ParticipantProfile`이 없다면 해당 요청은 유효한 요청이 됨
> - 요청 시 request body에 소속 대학(`university`)과 활성회원 여부(`isRegistered`)를 받음
>    + 값이 들어오지 않으면, default value(각각 `""`, `true`)로 DB에 저장됨
> - 이미 참여자인 사람이 또 요청하면 409로 응답

<br>
user id가 1인 진행자 User가 수강 권한을 요청하는 상황을 가정해보자.
이때의 request body는 아래 사진에 나와있는 것과 같이 구성하였다.
<img width="1206" alt="스크린샷 2022-10-31 오후 11 50 19" src="https://user-images.githubusercontent.com/90292371/199037049-c4754c56-b3bd-4105-974a-7b73bf91ecca.png">

### ☑️ DB Result
id가 1인 유저에게는 아직 수강 권한이 없었으므로 이 API 요청은 유효한 요청이라고 볼 수 있다.
그에 따라 이 유저에게 수강 권한이 성공적으로 부여되어, 아래 사진에서 제일 아래에 위치한 테이블이 **ParticipantProfile**인데 여기에 이 유저에 해당하는 row가 생성되었다. 
해당 row를 보면은, request body로 넘긴 `university`, `isRegistered` 값들로 잘 저장된 것을 확인해볼 수 있다.
<img width="1434" alt="스크린샷 2022-11-01 오전 3 51 39" src="https://user-images.githubusercontent.com/90292371/199086779-b298b548-82da-46ad-98eb-335dd563fc8f.png">

---
### 🚫 예외 케이스
#### 1) 이미 수강 권한 있는 진행자, 또다시 요청
: 앞서 수강 권한을 받은 User(id가 1인 유저 의미)가 또다시 수강 권한을 요청하는 API를 보낸다면 409 Conflict 에러
<img width="1206" alt="스크린샷 2022-10-31 오후 11 50 58" src="https://user-images.githubusercontent.com/90292371/199037201-e4ad440c-4c14-492e-9dde-428c5ea24f81.png">

#### 2) 수강생이 해당 요청을 보냄
: 수강생(여기서는 id가 3인 유저였음)은 이미 수강 권한을 가지고 있으므로 409 Conflict 에러
<img width="1162" alt="스크린샷 2022-10-31 오후 11 51 46" src="https://user-images.githubusercontent.com/90292371/199037382-fc24ede5-789b-40df-b2e8-4c3871295ba6.png">



---
# 🍒 세미나

### 🌵 세미나 생성
>  ### *구현 사항*
> - request body는 세미나 이름(name), 정원(capacity), 세미나 횟수(count), 정기 세미나 시간(time), 온라인 여부(online)를 포함해야 합니다. online 여부 외에는 하나라도 빠지면 400으로 응답하며, 적절한 에러 메시지를 포함합니다.
> - 요청을 보낸 사람이 세미나 진행자가 됨
>    + 즉, 이 요청을 보낸 유저가 INSTRUCTOR인지를 확인해야 함
> - 만약 이미 담당하고 있는 세미나가 있다면 더 맡을 수 없으로 추가 생성 불가

id가 1인 진행자 유저가 세미나를 생성하는 요청을 보냈다고 하자.

<img width="1206" alt="스크린샷 2022-10-31 오후 11 53 00" src="https://user-images.githubusercontent.com/90292371/199037683-ace012a5-3225-427e-9f66-24f9ea9b24b9.png">

### ☑️ DB Result
user id가 1인 유저가 세미나를 신규 개설했으므로
- **Seminars** 테이블에 API 요청시 함께 보낸 request body 정보를 토대로 생성된 세미나에 대한 row가 새롭게 하나 찍혀야 하고, 
- 또, **Users**와 **Seminars**의 매핑 테이블인 **UserSeminar** 테이블에도 row가 하나 찍혀야 한다.
아래 결과 사진을 보면, 예상했던 대로 각 테이블에 row가 추가된 것을 확인해볼 수 있다.

(위에서부터 순서대로 **UserSeminar**, **Users**, **Seminars** 테이블에 해당)
<img width="1549" alt="스크린샷 2022-11-01 오전 4 09 48" src="https://user-images.githubusercontent.com/90292371/199090203-6b5eba61-5617-4d20-915a-45c65192ba33.png">

---
### 🚫 예외 케이스
#### 1) 담당하고 있는 세미나가 있는 진행자, 또 다시 세미나 생성
: 앞서 세미나를 생성한 유저(id가 1인 유저)가 또 다시 세미나 생성 요청을 보냈다면, 400 Bad Request
<img width="1206" alt="스크린샷 2022-10-31 오후 11 53 30" src="https://user-images.githubusercontent.com/90292371/199037800-438bb10a-5cbc-4e7f-96c8-4404337ae420.png">
#### 2) 수강생, 세미나 생성 요청
: 403 Forbidden
<img width="1113" alt="스크린샷 2022-11-01 오전 4 12 43" src="https://user-images.githubusercontent.com/90292371/199090682-905bbcd1-97c4-4598-9c35-7927730c8122.png">


---
### ✋ 잠깐 ! 넘어가기 전에 !!
앞으로 남은 API들의 구현 결과를 더 잘 드러내기 위해서는, 추가적인 유저와 세미나가 있어야 할 것 같더라구요.
그래서 아래 사진과 같이 진행자, 참여자, 세미나를 추가해 넣었습니다.
<img width="1390" alt="스크린샷 2022-11-01 오후 1 48 43" src="https://user-images.githubusercontent.com/90292371/199161220-73fce29d-d095-4ffc-b963-ec28d372fd02.png">
<img width="1340" alt="스크린샷 2022-11-01 오후 2 39 55" src="https://user-images.githubusercontent.com/90292371/199166605-ae5504cb-219c-43f0-85e4-f8c6d38fbd2c.png">

요기서부터는 이러한 DB 상태에서 API 요청이 이루어진 걸로 이해해주시면 감사하겠습니다 🙂


--
### 🌵 세미나 수정
>  ### *구현 사항*
> - request body에 수정하고자 하는 seminar id가 포함될 수도 있고 아닐 수도 있음
      + 만약 seminar id가 없다면 요청을 보낸 유저가 담당하는 세미나의 정보를 수정하는 걸로 이해
> - 요청을 보낸 사람이 세미나 진행자가 됨
>    + 즉, 이 요청을 보낸 유저가 INSTRUCTOR인지를 확인해야 함
> - 만약 이미 담당하고 있는 세미나가 있다면 더 맡을 수 없으로 추가 생성 불가

user id가 1인 진행자 유저가 본인이 담당하고 있는 세미나명을 수정하고자 함.
<img width="1206" alt="스크린샷 2022-10-31 오후 11 54 08" src="https://user-images.githubusercontent.com/90292371/199037948-fd3b439b-c3a0-46b0-bddf-2836f5e0fd6c.png">

---
### 🚫 예외 케이스
#### 1) 진행자 - 본인이 맡은 세미나가 아닌 세미나 정보를 수정
: user id가 1인 진행자 유저는 seminar id가 1인 세미나를 맡고 있다. 즉, seminar id가 2인 세미나의 담당자는 아니므로, 아래와 같이 수정 요청을 보내게 되면 403 Forbidden 에러
<img width="1206" alt="스크린샷 2022-10-31 오후 11 54 42" src="https://user-images.githubusercontent.com/90292371/199038076-f4cad260-fcfc-4de5-b4d3-c17e6ca5204f.png">

#### 2) 수강생의 세미나 수정 요청
: 수강생에게는 세미나 수정 권한이 없으므로 403 Forbidden 에러
아래는 user id가 3인 수강생 유저가 seminar id가 1인 세미나의 정보를 수정하고자 하는 요청을 보냈을 때의 결과 사진이다.
<img width="1162" alt="스크린샷 2022-10-31 오후 11 55 31" src="https://user-images.githubusercontent.com/90292371/199038321-f671da5a-f5b6-4e96-a306-719aba2c8d1e.png">


---
### 🌵 세미나 조회
#### ✔️ By `id`
- id가 1인 세미나 조회
<img width="1206" alt="스크린샷 2022-10-31 오후 11 56 04" src="https://user-images.githubusercontent.com/90292371/199038471-9dca8c73-6d75-48b1-a919-e081fd452064.png">
- id가 3인 세미나 조회
<img width="1113" alt="스크린샷 2022-11-01 오후 1 52 36" src="https://user-images.githubusercontent.com/90292371/199161648-382ea727-0ebb-487d-a982-1026c9a40c59.png">

#### ✔️ By `name`
- name 파라미터를 입력하지 않으면 세미나 전체 조회
<img width="1304" alt="스크린샷 2022-11-01 오후 2 19 39" src="https://user-images.githubusercontent.com/90292371/199164366-8f1ef081-90be-4afa-a24d-eca32d9675a2.png">

- "클래스"를 포함하는 세미나 조회
: seminar id가 2, 3인 세미나가 조회되어야 한다.
<img width="1304" alt="스크린샷 2022-11-01 오후 2 20 28" src="https://user-images.githubusercontent.com/90292371/199164458-63ce560a-78c8-4954-a786-e0b07116606f.png">
이때, "천변 클래스(seminar id = 3)"에 "백마리(user id = 9)"가 참여중이긴 하지만 이 수강생의 `isActive` 값이 false라 참여자수를 셀 때에는 포함되면 안 된다. 즉, 위 사진에서와 같이 `participantCount`는 0으로 나오는 게 맞음. 또한, "갓기 표정 클래스"에는 현재 "수강생2(id = 6)"과 "수강생3(id = 7)"이 참여중이므로 `participantCount` 값은 2. <br>
그리고, `order`에 대한 값을 입력하지 않았으므로 세미나 최근 생성 기준으로 response body가 구성되어야 한다.

#### ✔️ By `name` + 정렬
바로 앞 케이스에서와 동일한 요청을 보내되, `order` 파라미터에 "earliest"라는 값을 보내어 response body가 세미나 생성 일자 기준으로 내림차순으로 정렬되게끔 하였다.
<img width="1304" alt="스크린샷 2022-11-01 오후 2 20 57" src="https://user-images.githubusercontent.com/90292371/199164512-f953e237-df45-41fc-9991-17ae0a625aac.png">


---
### 🌵 세미나 참여 (PARTICIPANT)
"수강생1(id = 3)"이 "셈이나(id=1)" 강좌를 수강
<img width="1129" alt="스크린샷 2022-11-01 오후 2 22 33" src="https://user-images.githubusercontent.com/90292371/199164663-f7355cf0-f9d5-41cb-b60d-9782b997a885.png">

### ☑️ DB Result
위 정보가 "UserSeminar" 테이블에도 저장되었음을 확인(마지막 row)
<img width="1313" alt="스크린샷 2022-11-01 오후 2 28 30" src="https://user-images.githubusercontent.com/90292371/199165277-04a7c48f-d0d8-40c8-a285-1277214b35bf.png">

---
### 🚫 예외 케이스
#### 1) 정원 초과
"수강생4(user id = 10)"가 "셈이나(id=1)"을 수강하려고 하나, 이미 해당 강좌는 정원이 1명이고 또 "수강생1(user id=3)"이  세미나 1에  참여 중이므로 정원이 가득찬 상태이다. 
즉, 정원 초과로 400 Bad Request 에러

<img width="1065" alt="스크린샷 2022-11-01 오후 2 31 35" src="https://user-images.githubusercontent.com/90292371/199165589-e8d64477-ab8d-4a42-9527-0472a17f1e40.png">

---
### 🌵 세미나 참여 (INSTRUCTOR)
"보성(user id = 2)"은 현재 담당하고 있는 세미나가 없음.
따라서 다른  강좌에 진행자 권한으로 참여 가능.
아래는 해당 유저가 "갓기 표정 클래스(seminar id = 2)"의 참여 요청을 보낸 결과에 해당한다.
<img width="1398" alt="스크린샷 2022-11-01 오후 2 25 18" src="https://user-images.githubusercontent.com/90292371/199164956-6818eb23-262c-4eb8-b591-2341436c9201.png">

### ☑️ DB Result
<img width="1313" alt="스크린샷 2022-11-01 오후 2 28 46" src="https://user-images.githubusercontent.com/90292371/199165312-386a92fe-9c2c-4090-9a73-f6ef355374b6.png">


---
### 🌵 세미나 드랍
"수강생3(user id = 7)"이 본인이 수강하는 "갓기 표정 클래스(id=2)"를 드랍하고자 한다.
해당 요청의 response body를 보면은, 우리의 기대대로 이 유저의 `isActive` 값이 false로 나오고 있다.
<img width="1446" alt="스크린샷 2022-11-01 오후 2 33 35" src="https://user-images.githubusercontent.com/90292371/199165805-2fb9db8c-213f-4d91-a673-c0ca278e47ca.png">

### ☑️ DB Result
**UserSeminar** 테이블에도 반영되었음을 확인해볼 수 있다. (id = 6인 row 참고)
<img width="1313" alt="스크린샷 2022-11-01 오후 2 34 05" src="https://user-images.githubusercontent.com/90292371/199165870-a537e20d-58aa-4628-9624-e0a35e8f38ed.png">

### 🚫 예외 케이스
#### 1) 진행자 드랍 불가
: 천지훈(user id = 8)이 본인의 "천변 클래스" 세미나(id=3)를 드랍하려고 하나 불가능. 403 Forbidden 에러
<img width="1021" alt="스크린샷 2022-11-01 오후 2 36 08" src="https://user-images.githubusercontent.com/90292371/199166112-75ab4059-1894-42f6-92fa-c8bfe1232bd2.png">

---
### 🌵 세미나 드랍 후 재수강 불가
앞서 "갓기 표정 클래스(id = 2)"를 드랍했던 "수강생3(user id = 7)"이 마음을 바꾸어 다시 수강을 요청했다고 해보자. 슬프게도 드랍 후 재수강은 불가능하므로, 400 Bad Request 에러
<img width="1021" alt="스크린샷 2022-11-01 오후 2 34 55" src="https://user-images.githubusercontent.com/90292371/199165989-ea708b9b-67bb-4666-8445-9743cbbd1713.png">


---
# 🍒 AOP 요청별 시간 로깅
<img width="1157" alt="스크린샷 2022-11-01 오전 4 17 02" src="https://user-images.githubusercontent.com/90292371/199091406-93715f1f-5fec-4ccb-b011-371688a5e248.png">

<img width="1200" alt="스크린샷 2022-11-01 오전 4 16 26" src="https://user-images.githubusercontent.com/90292371/199091335-1fb34b0d-bcb3-402b-ad61-ebb76e8a82a2.png">
<img width="1356" alt="스크린샷 2022-11-01 오전 12 26 09" src="https://user-images.githubusercontent.com/90292371/199045366-4ea2446d-1e59-46ce-9d75-11e3818e88ea.png">
<img width="1337" alt="스크린샷 2022-11-01 오전 12 25 54" src="https://user-images.githubusercontent.com/90292371/199045311-8d47a92d-fce2-4801-bc46-10190b49bff0.png">
<img width="1350" alt="스크린샷 2022-11-01 오전 12 25 38" src="https://user-images.githubusercontent.com/90292371/199045240-11ac4874-f8b8-42f9-a687-b40b006cbdb8.png">

이처럼 모든 요청이 (현재 데이터 크기 상으론..) 전부 2s 안에 수행됨을 확인

Jhvictor4 committed:
Update README.md

Boseong-Seo committed:
application.yaml 수정
- DB의 timezone을 Asia/Seoul로 설정
- jwtSecret 값을 적절히 추가

Boseong-Seo committed:
DB 연결 및 설정(dependencies 등) 수정

Boseong-Seo committed:
회원가입, 로그인, JWT 인증 구현

Boseong-Seo committed:
테이블 생성을 위한 entity 설정
- 테이블간 연관관계를 고려하여 적절히 매핑

Boseong-Seo committed:
유저 프로필 읽기 및 수정 기능 구현
- 관련 엔티티의 레퍼지토리 생성
- 제시한 json spec에 맞는 response body를 리턴할 수 있도록, querydsl 추가
   - querydsl을 사용하기 위한 custom repository 생성

Boseong-Seo committed:
세미나 관련 기능 구현
- 세미나 생성, 불러오기, 수정, 드랍
- 각 request마다 내보내는 response 양식을 준수하기 위한 DTO 클래스 생성
  - QueryDSL Projection을 이용하여 Entity-DTO 매핑

Boseong-Seo committed:
세미나 참여, 진행 API 요청시 필요한 request body

Boseong-Seo committed:
진행자, 수강 권한 요청시 필요한 request body

Boseong-Seo committed:
예외 처리

Boseong-Seo committed:
요청 별 시간 로깅

Boseong-Seo committed:
exception은 하나의 enum class로 관리하고자
따라서, 해당 파일이 필요없다고 판단해 삭제 처리

Boseong-Seo committed:
테스트 파일
+) 그 외 미처 놓치고 제때 커밋 못한 파일들

Boseong-Seo committed:
프로필 수정 요청 API 수정
- request body 수정
  + role 필드 삭제
  + 유효성 검사 추가 : Controller에서 @Valid annotation 누락 -> 추가
- 한 유저가 진행자이면서도 참여자일 수 있는 경우 고려

Boseong-Seo committed:
프로필 수정 요청 API 수정
- request body 수정
  + role 필드 삭제시 함께 삭제해야 했던 유효성 내용 삭제 누락

Boseong-Seo committed:
- 네이밍 변경
  + 각 함수 및 클래스/dto가 어떤 역할을 하는지 이름에서 보다 명확히 드러내기 위해 네이밍 수정
- Return type 수정
  + Any 타입 -> 보다 명확하게 수정

Boseong-Seo committed:
- 예외 처리 보완
  + 세미나 드랍 요청(SeminarService의 `dropSeminar` 함수) 관련 예외 (수강 이력 없음에도 드랍 요청)
- 예외 케이스 네이밍 수정
  + 어떤 예외 케이스에 해당하는지 보다 명확히 나타내기 위해 네이밍 수정

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/user/api/request/SignUpRequest.kt:
```diff
@@ -1,7 +1,27 @@
+    @field:NotBlank(message = "비밀번호는 필수 입력값입니다.")
+    val password: String?,
+    @field:NotNull(message = "회원유형을 선택해주세요.")
+    @field:Pattern(
```
이렇게 Request body validation 할 시 정규식 패턴을 이용해서 바로 검사할 수 있다는 사실을 처음 알아가네요! 사전에 Body를 받으면서 양식에 맞지 않는 것을 빠르게 controller 단에서 예외처리 할 수 있다는 것이 좋은 거 같습니다!

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/config/WebConfig.kt:
```diff
@@ -1,13 +1,73 @@
+}
+
+@Configuration
+class AuthInterceptor(
```
저 같은 경우는 TokenVerifyInterceptor를 이용해서 구현했었는데, AuthInterceptor를 이용한 방식도 찾아봐야겠습니다. 좋은 참고가 되었습니다!

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/user/domain/enums/RoleType.kt:
```diff
@@ -0,0 +1,5 @@
+package com.wafflestudio.seminar.core.user.domain.enums
+
+enum class RoleType {
+    PARTICIPANT, INSTRUCTOR
```
enum을 이용한 role 처리를 하셔서 깔끔하고 코드 작성시 실수할 가능성도 줄 거 같습니다. 배워야 할 점 같습니다!!

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/user/service/AuthTokenService.kt:
```diff
@@ -1,47 +1,82 @@
+        throw SeminarException(ErrorCode.EMAIL_NOT_FOUND)
+      } else return email
+      
+    } catch (e: SignatureException) {       // 서명 key가 잘못되었을 때
+      throw SeminarException(ErrorCode.FORBIDDEN)
+    } catch (e: IllegalArgumentException) { // claimJws string이 null일 때
+      throw SeminarException(ErrorCode.FORBIDDEN)
+    } catch (e: ExpiredJwtException) {      // 토큰 유효 시간 넘겼을 때
+      throw SeminarException(ErrorCode.EXPIRED_TOKEN)
+    }
```
전 authentication에서 오류 처리를 어떻게 해야 되는지 감을 잘 잡지 못해서 일일이 controller단에서  ResponseEntity로 unauthorized를 보내도록 처리했었는데, AuthTokenService에서 처리하는 것이 훨씬 중복되는 코드도 줄이고 깔끔하겠네요! 배울 점인거 같습니다.

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/user/service/AuthTokenService.kt:
```diff
@@ -1,47 +1,82 @@
+    val prefixRemoved: String = removePrefix(authToken)
+    
+    try {
+      val claim: Jws<Claims> = Jwts.parserBuilder()
+        .setSigningKey(signingKey)
+        .build()
+        .parseClaimsJws(prefixRemoved)
```
이 부분은 별개의 method로 분리하는게 좋아보여요. 지금은 getCurrentUserId에서만 쓰이지만, 나중에 추가적으로 token에 정보가 추가되고, 그 정보를 가져오는 기능을 implement해야 될 때 이 부분이 공통적으로 사용될 것이기 때문이에요. 추가적으로 코드가 깔끔해보이기도 하고요..!!

Boseong-Seo replied:
네네 조언 감사합니다! 안그래도 혁준님 코드에서는 이 부분이 별도의 method로 분리되어 있는 것을 보며 배워야 할 점이라고 생각했거든요. 작은 단위이더라도 별도의 method로 분리하는 습관(?)을 들여봐야 겠어요

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/common/BaseTimeEntity.kt:
```diff
@@ -1,26 +1,29 @@
     @CreatedDate
     @Column(columnDefinition = "datetime(6) default '1999-01-01'")
-    var createdAt: LocalDateTime? = null
+    var createdAt: LocalDateTime? = LocalDateTime.parse(
+        LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")),
+        DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
+    )
 
-    @CreatedDate
+    @LastModifiedDate
     @Column(columnDefinition = "datetime(6) default '1999-01-01'")
-    var modifiedAt: LocalDateTime? = null
+    var modifiedAt: LocalDateTime? = LocalDateTime.parse(
+        LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")),
+        DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
+    )
```
@CreateDate와 @LastModifedDate가 있으면 생성/수정시에 알아서 해당 시간을 db로 저장해주는거 아니었나요?? default 값을 null이 아닌 추가로 설정해주신 이유가 궁금하네요...!!

Boseong-Seo replied:
작성해주신 대로 **@Createdate**와 **@LastModifiedDate**가 생성/수정 시의 시각을 알아서 DB에 저장은 해주나, 저장된 시각의 포맷이 millisec까지 표기가 되는 형태(e.g. `2022-11-22 09:48:50.1234`)더라고요..! 그래서 저는 초 단위까지만 나타내는 `yyyy-MM-dd HH:mm:ss` 포맷으로 DB에 저장됐음 해서, 위같이 코드를 작성했습니다.
그런데 이렇게 하면은 위 annotation을 쓰는 의미가 없는 것 같기도 하네요.. 위 방식 외에 저장되는 포맷 방식을 변경할 수 있는 방식이 있는지 구글링해봐야 겠습니다 😵‍💫

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/UserSeminar/domain/UserSeminarEntity.kt:
```diff
@@ -0,0 +1,47 @@
+@Table(name="UserSeminar")
+@EntityListeners(AuditingEntityListener::class)
+class UserSeminarEntity (
+    @JsonBackReference
```
순환참조를 방지하는데 이런 방법도 있었군요! 참고하겠습니다!!

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/user/repository/UserCustomRepository.kt:
```diff
@@ -0,0 +1,103 @@
+                            seminarEntity.id, seminarEntity.name,
+                            userSeminarEntity.joinedAt,
+                            userSeminarEntity.isActive,
+                            userSeminarEntity.droppedAt?: null
```
?: null은 삭제해도 될 거 같네요.

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/user/service/UserService.kt:
```diff
@@ -0,0 +1,144 @@
+            }
+            
+            lateinit var newUser: UserEntity
+            when (request.role) {
+                "PARTICIPANT" -> {
+                    newUser = UserEntity(this.email, this.username!!, encoder.encode(this.password), RoleType.PARTICIPANT)
+                    newUser.participantProfile =
+                        ParticipantProfile(this.university?:"", this.isRegistered?:true, newUser)
+                }
+                "INSTRUCTOR" -> {
+                    newUser = UserEntity(this.email, this.username!!, encoder.encode(this.password), RoleType.INSTRUCTOR)
+                    newUser.instructorProfile =
+                        InstructorProfile(this.company?:"", this.year, newUser)
+                }
+            }
```
Role에 대한 Enum Class를 정의하신 만큼, Enum class의 name을 이용하여 매칭하는 것이 하드코딩하는 방식보다 더 좋아보입니다...!

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/user/service/UserService.kt:
```diff
@@ -0,0 +1,144 @@
+            throw SeminarException(ErrorCode.NOT_AUTHORIZED)
+        } else {
+            // 로그인 성공 - 마지막 로그인 시각 업데이트
+            userEntity.lastLogin = LocalDateTime.parse(
+                LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")),
+                DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
+            )
```
LocalDateTime을 정의하는데 LocalDateTime을 string으로 포맷한 다음, 이를 다시 parse하는 방식으로 구현하신 이유가 있나요?? 궁금합니다!

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/user/service/UserService.kt:
```diff
@@ -0,0 +1,144 @@
+    override fun getUserProfile(userId: Long, targetUserId: Long): UserDTO {
+        return with(userRepository) {
+            // 확인하려는 유저 정보가 DB 상에 존재하는가?
+            if(this.findById(targetUserId).isEmpty)
+                throw SeminarException(ErrorCode.USER_NOT_FOUND)
```
이런 부분에서 저는 주로 findByIdOrNull로 해서 Optional type보다는 ? 타입으로 받은 후에 처리해요. 이건 사실 스타일 차이이긴 한데, 저는 그게 더 깔끔하다고 느껴지더라고요! 그냥 이렇게 하는 사람도 있다... 라고만 참고해주시면 좋겠습니다 !!

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/user/service/UserService.kt:
```diff
@@ -0,0 +1,144 @@
+                    "PARTICIPANT" -> RoleType.PARTICIPANT
+                    "INSTRUCTOR" -> RoleType.INSTRUCTOR
+                    else -> targetUser.role
+                    }
```
이 부분도 앞에서 request body validation을 통해 이미 잘못된 문자열이 들어올 경우는 없으니까,
RoleType.valueOf(this.role) 이런식으로 해서 바로 enum class type으로 변환시키는게 더 깔끔할 거 같습니다.

Boseong-Seo replied:
오 네넵 조언 감사합니다 !!

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/UserSeminar/repository/UserSeminarCustomRepository.kt:
```diff
@@ -0,0 +1,26 @@
+package com.wafflestudio.seminar.core.UserSeminar.repository
+
+import com.wafflestudio.seminar.common.QueryDslConfig
+import com.wafflestudio.seminar.core.UserSeminar.domain.QUserSeminarEntity.userSeminarEntity
+import com.wafflestudio.seminar.core.user.domain.enums.RoleType
+
+interface UserSeminarCustomRepository {
+    fun checkInstructingSeminars(userId: Long): Long
+}
+
+class UserSeminarCustomRepositoryImpl (
+    private val queryDslConfig: QueryDslConfig
+) : UserSeminarCustomRepository{
+    private val queryFactory = queryDslConfig.jpaQueryFactory()
+    
+    override fun checkInstructingSeminars (userId: Long): Long {
+        return queryFactory
+            .select(userSeminarEntity.id.count())
+            .from(userSeminarEntity)
+            .where(
+                userSeminarEntity.user.id.eq(userId)
+                .and(userSeminarEntity.role.eq(RoleType.INSTRUCTOR))
+                .and(userSeminarEntity.isActive.eq(true)))
+            .fetchOne()!!
+        }
+}
```
이 부분은 기능상 묶이는 UserSeminarRepository.kt에 같이 작성하는 것도 괜찮을거 같습니다...?!! 물론 스타일 차이이긴 합니다.

huGgW commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/common/SeminarExceptionHandler.kt:
```diff
@@ -1,10 +1,90 @@
+    }
     
-    // TODO as-you-wish
+    @ExceptionHandler(value = [MethodArgumentNotValidException::class])
+    fun handler(e: MethodArgumentNotValidException): ResponseEntity<ErrorResponseList> {
+        val request : HttpServletRequest =
+            (RequestContextHolder.getRequestAttributes() as ServletRequestAttributes?)!!.request
+        
+        var errorList: List<ErrorReason> = 
+            e.bindingResult.allErrors.map {
+                ErrorReason(
+                    field= (it as FieldError).field,
+                    value= (it as? FieldError)?.rejectedValue,
+                    reason= it.defaultMessage!!
+                )
+            }.toList()
+        return ResponseEntity(
+            ErrorResponseList(
+                status= 400,
+                errors= errorList,
+                path= request.requestURI
+            ),
+            HttpStatus.BAD_REQUEST
+        )
+    }
```
제가 Request Body에 대한 Exception을 handle할 때 MethodArtumentNotValidException 말고도 다른 Exception이 호출되는 경우도 있었어서 결국 직접 Controller마다 bindingresult를 정의해서 처리를 했었는데, 그런 경우는 없으셨었나요? 이렇게 코드 구성하신 것을 보니 제가 그때 뭔가 잘못 코드를 작성했을수도 있겠다는 생각이 드네요.

huGgW commented:
우선 늦은 리뷰 죄송합니다.....!! 🙏
전체적으로 커밋도 세부적으로 해주시고, 주석도 꼼꼼하게 달아주셔서 정말 리뷰하기 편했어요...!!
리뷰하기는 했지만, 사실 자잘한 것들에 대한 피드백이 있고 대부분은 제가 배우거나 질문하는 내용이었어요..!
특히 QueryDsl, Authentication 처리와 Exception Handling에 관한 부분이 매우 인상깊었습니다.
저도 자극 받아서 더 열심히 해야되겠어요!
수고하셨어요!!

huGgW committed:
Added HibernateQueryCounter

huGgW committed:
Added helper for creating entities

huGgW committed:
Added methods to helper

huGgW committed:
Test for makeSeminar method

huGgW committed:
Refactoring

huGgW committed:
Fixed createUserSeminarEnitity

huGgW committed:
Added test for editSeminar method

huGgW committed:
Added test for findSeminarContainingWord

huGgW committed:
Fixed some n+1 testing, and added test for findSeminarById

Boseong-Seo commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/UserSeminar/domain/UserSeminarEntity.kt:
```diff
@@ -0,0 +1,47 @@
+    )
+    var droppedAt: LocalDateTime? = null
+
+    @Enumerated(EnumType.STRING)
+    var role: RoleType = user.role
```
코멘트 남겨주신 것처럼, 저도 Enum type 설정 시 데이터베이스 용량 관련해서 고민을 했었는데요.
하지만, 이 프로젝트에서 저희가 다룰 데이터의 크기가 그렇게 크진 않을 거란 점에서, DB의 용량보다는 사용자 측면을 더 중시해도 되겠다고 판단했습니다. 다시 말해, 유저가 이 테이블을 봤을 때 `role` 필드에 담긴 값이 1, 2로 되어 있는 것보다는 "instructor", "participant"로 되어 있는 게 이해가 단번에 되는 건 사실이니까요. (디버깅 시 ORDINAL로 되어 있어야 제가 편리했던 측면도 있었고요 ㅎ ... )

Jhvictor4 replied:
좋은 포인트인데요, ORDINAL은 EnumType이 추가되거나, 변경 (특히 삭제) 될 때 큰 리스크를 가지고 있어요..!
STRING으로 해도 1글자당 4바이트 수준인 점을 감안하면 보통은 STRING 형태 그대로 저장하는 편인 것 같습니다. :)

Boseong-Seo commented on seminar2/src/main/kotlin/com/wafflestudio/seminar/core/seminar/api/SeminarController.kt:
```diff
@@ -0,0 +1,73 @@
+    ): Any? = seminarService.regSeminar(userId, seminarId, request)
+
+    
+    @Transactional
+    @Authenticated
+    @DeleteMapping("/api/v1/seminar/{seminar_id}/user")
+    fun dropSeminar(
+        @RequestHeader("Authorization") accessToken: String,
+        @UserContext userId: Long,
+        @PathVariable("seminar_id") seminarId: Long,
+    ): SeminarDTO = seminarService.dropSeminar(userId, seminarId)
+}
```
의미가 있었던 것은 아니고 .. controller 단에서 transactional annotation을 붙인 것으로 충분하다고 생각했을 뿐이었습니다. 😢 Service 단에서도 보다 작은 함수 단위로 안전한(?) roll back을 위한 **@Transactional**을 달았어야 했던 게 아닌가 싶네요

huGgW committed:
Added exception http status code checking

huGgW committed:
Added test for registerSeminar

Boseong-Seo committed:
- 세미나 드랍 요구사항 반영
  + 세미나 드랍 method에서, 해당 세미나에 참여하고 있지 않은 수강생이 드랍 요청을 했다면 200으로 응답하도록 수정 (이전 버전에선 400 에러 처리)

Boseong-Seo committed:
role 필드에 대한 validation 수정
- **@EnumPattern** annotation 생성
  + javax validation custom enum constrains

Boseong-Seo committed:
Merge pull request from huGgW/seminar3-testing

우혁준 Seminar3 과제

Boseong-Seo committed:
- RegisterRequest의 role field type 변경에 따른 수정
  : Enum -> String으로 수정
- 테스트를 할 클래스를 알아서 잘 찾도록 파일명 및 테스트 클래스명 수정

Boseong-Seo committed:
[FIX] Entity 속성 수정
- user 삭제 시 그와 연결된 profile이 삭제되지 않는 문제 수정 (orphanRemoval 설정)

Boseong-Seo committed:
JPA N+1 문제 해결
- 그를 위한 entity class들 리팩토링

Jhvictor4 commented:
💯💯💯

Boseong-Seo committed:
디버깅 목적으로 작성했던 print 구문 삭제

Boseong-Seo committed:
테스트 시 문제 생기는 부분 수정

Boseong-Seo committed:
[Fix] 코드 리팩토링
- 그에 따라 테스트 코드도 일부 수정

Boseong-Seo committed:
[Fix] JPA N+1 문제 해결
- 테스트 코드 전부 통과 확인!
